### DISCUSSION TITLE
How OpenCL is Different from gpgpu-rs

### DISCUSSION THREAD

**Discussion by RahulDas-dev** [UpsettingBoy/gpgpu-rs#11] [Q&A] [answered]
Hi, 

I was working on a gaussian filter for images using GPU, I was using [OpenCL](https://github.com/kenba/opencl3), But I stumbled across your Library, however. I know very little about WGPU. So, I Have Couple of question on my mind.

1. If I write the same code using OpenCL and gpgpu-rs, will both run on the same hardware/GPU?
2. OpenCL requires the OpenCL driver to be installed on the machine. Does GPGPU require any driver?
3. Is GPGPU capable of processing matrix and linear algebra operations? if yes, then is there any supporting Library for it.
4. What is difference between these two technologies? is there any blog/Doc describing those technologies in details?

**Comment by jeronimosg** [comment, marked as answer]
Let me answer you per question :). I think you are mixing some terms, I recommend you to start reading 4. and the paragraph bellow.
1. More or less. OpenCL can be executed on accelerators other than GPUs such as FPGAs, CPUs etc. gpgpu-rs only runs on GPUs since it's using wgpu as backend, which use dx12, metal and vulkan. The main problem is that OpenCL GPU programs, writen in OpenCL C, cannot be directly executed with any of the graphics backends mentioned above (vulkan, ...) because they are different. IIRC Google is working on a [OpenCL C -> SpirV compiler](https://github.com/google/clspv).
2. You are correct. OpenCL requires a driver (sometimes not easy to install). Using graphics only requires the driver of the GPU, so using gpgpu-rs is very straightforward in terms on installing dependencies.
3. If you are looking into matrix multiplication gpgpu-rs provides integration with ndarray in rust (I believe that it needs more work :-) ). For the kind of workload you are describing I think CUDA with CuBLAS is a great combo.
4. With that I think you are confused about what GPGPU means. GPGPU stands for General Purpose GPU, which means using the GPU for computing workloads other than graphics. OpenCL is a GPGPU framework that runs on almost any hardware (if the vendor provides driver). CUDA is also  a GPGPU framework (only NVIDIA's). You can search the differences between OpenCL and CUDA or among any other GPGPU framework that exists.

gpgpu-rs is a rust GPGPU framework built upon wgpu, which is an abstraction layer over the different graphics backends that exists (vulkan, etc) using the compute capabilities of this graphics APIs to do GPGPU.

**Reply by RahulDas-dev** [reply]
Thanks for enlightening me. I will definitely replace my OpenCL codebase with WGPU
